cloud-energy-collection/iriscasttools/utils.py

```diff
@@ -0,0 +1,208 @@
+    """Run a bash command with given arguments and return output
+
+    Keyword Arguments
+        cmd_args: a string representing bash command to run
+    """
+    with subprocess.Popen(
+        cmd_args, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+    ) as out:
+        # assert we did not find any errors
+        assert not out.stderr.read().decode()
+        res_out = out.stdout.read().decode()
+    return res_out
+
+
+def check_ipmi_conn():
+    """Check if IPMItool exists and can be connected to
+
+    Checks if device exists at any of these locations /dev/ipmi0, /dev/ipmi/0 or /dev/ipmidev/0
+    which imply that ipmi-dcmi can be used to get power info
+    """
+    return any(
+        Path(f).exists() for f in ["/dev/ipmi0", "/dev/ipmi/0", "/dev/ipmidev/0"]
+    )
+
+
+def ipmi_raw_power_query():
+    """
+    Get raw power query from IPMItool
+
+    Calls ipmi-dcmi command to get all power statistics
+    """
+    try:
+        return retry(
+            lambda: run_cmd("/usr/sbin/ipmi-dcmi --get-system-power-statistics"),
+            AssertionError,
+            retries=3,
+            delay=3,
+            backoff=2,
+        )
+
+    except AssertionError:
+        return None
+
+
+def to_csv(stats: Dict, include_header: bool = False):
+    """convert dictionary into csv string
+
+    Keyword arguments
+    stats -- dict, key-value pairs to be returned in csv format
+    include_header, bool, flag to set if dictionary keys should be returned as header
+    """
+    res = ",".join([str(s) for s in stats.values()])
+    if include_header:
+        return f"{','.join(stats.keys())}\n{res}"
+    return res
+
+
+def get_ipmi_power_stats(*args):
+    """Get ipmi power stats as dictionary.
+
+    Keyword arguments
+    *args -- a set of fields to parse and collect.
+    can be one or more of:
+        "current_power": instantaneous reading in Watts,
+        "minimum_power_over_sampling_duration: minimum Watt reading from last sample till now
+        "maximum_power_over_sampling_duration": maximum Watt reading from last sample till now
+        "average_power_over_sampling_duration": average Watt reading from last sample till now
+        "statistics_reporting_time_period": sample period in milliseconds
+        "time_stamp": current timestamp for reading taken
+        "power_measurement": whether power measurement is active on host
+    """
+    if not check_ipmi_conn():
+        return None
+    res = ipmi_raw_power_query()
+
+    power_stats = {x: "" for x in args}
+
+    if res:
+        for line in res.splitlines():
+            line_str = line.split(":")
+            stat = line_str[0].strip().lower().replace(" ", "_")
+            raw_val = ":".join(line_str[1:]).strip()
+
+            if stat in power_stats.keys():
+                if stat == "time_stamp":
+                    stat_val = re.search(
+                        r"\d{2}/\d{2}/\d{4}\s-\s\d{2}:\d{2}:\d{2}", raw_val
+                    ).group(0)
```

Comment:
When regex is involved, we should have line-by-line comments explaining what you are doing, as it can quickly suck up dev time

Ideally, I would push you to simplify with tools like `str.split()` then filter using lists instead of regex. It's more terse, but always way easier to maintain or change long term